symbolic-il2cpp/src/usym.rs

```diff
@@ -0,0 +1,227 @@
+//! Parser for the UsymLite format.
+//!
+//! This format can map il2cpp instruction addresses to managed file names and line numbers.
+//!
+//! Current state: This can parse the UsymLite format, a method to get the source location
+//! based on the native instruction poiter does not yet exist.
+
+use std::borrow::Cow;
+use std::mem;
+use std::ptr;
+
+use anyhow::{Error, Result};
+
+/// The header of the usym file format.
+#[derive(Debug, Clone)]
+#[repr(C)]
+struct UsymHeader {
+    /// Magic number identifying the file, `b"usym"`.  Probably.
+    magic: u32,
+
+    /// Version of the usym file format.  Probably.
+    version: u32,
+
+    /// Number of [`UsymRecord`] entries.
+    ///
+    /// These follow right after the header, and after them is the string table.
+    record_count: u32,
+
+    /// UUID of the assembly(?).
+    id: u32,
+
+    /// Name of something, the "assembly"?  Offset into string table.
+    name: u32,
+
+    /// Name of OS.  Offset into string table.
+    os: u32,
+
+    /// Name of architecture.  Offset into string table.
+    arch: u32,
+}
+
+/// A record mapping an IL2CPP instruction address to managed code location.
+///
+/// This is the raw record as it appears in the file, see [`UsymRecord`] for a record with
+/// the names resolved.
+#[derive(Debug, Clone, Copy)]
+#[repr(C, packed)]
+struct UsymRawSourceRecord {
+    /// Instruction pointer address, relative to ...?
+    address: u64,
+    /// Managed symbol name as offset in string table.
+    symbol: u32,
+    /// Reference to the managed source file name in the string table.
+    file: u32,
+    /// Managed line number.
+    line: u32,
+    // These might not even be u64, it's just 128 bits we don't know.
+    _unknown0: u64,
+    _unknown1: u64,
+}
+
+/// A record mapping an IL2CPP instruction address to managed code location.
+#[derive(Debug, Clone)]
+struct UsymSourceRecord<'a> {
+    /// Instruction pointer address, relative to ...?
+    address: u64,
+    /// Symbol name of the managed code.
+    symbol: Cow<'a, str>,
+    /// File name of the managed code.
+    file: Cow<'a, str>,
+    /// Line number of the managed code.
+    line: u32,
+}
+
+/// Reader for the usym symbols file format.
+pub struct UsymSymbols<'a> {
+    /// File header.
+    header: &'a UsymHeader,
+    /// Instruction address to managed code mapping records.
+    records: &'a [UsymRawSourceRecord],
+    /// The string table.
+    ///
+    /// This is a large slice of bytes with length-prefixed strings.  The length is a
+    /// little-Indian u16.
```

Comment:
Awwwww this just made my day :-D I was wondering when this typo is gonna happen :-D